common/thorhelper/thorread.cpp

```diff
@@ -0,0 +1,825 @@
+
+
+protected:
+    offset_t getLocalOffset();
+
+protected:
+    Owned<ISerialStream> input;
+    Owned<IFileIO> inputfileio;
+    CThorContiguousRowBuffer inputBuffer;
+    Owned<IEngineRowAllocator> outputAllocator;
+    IOutputMetaData * expectedDiskMeta = nullptr;
+    IOutputMetaData * projectedDiskMeta = nullptr;
+    IOutputMetaData * actualDiskMeta = nullptr;
+    unsigned expectedCrc;
+    unsigned projectedCrc;
+    unsigned actualCrc;
+    MemoryBuffer encryptionKey;
+    size32_t readBufferSize = defaultReadBufferSize;
+    bool grouped = false;
+    bool stranded = false;
+    bool compressed = false;
+    bool blockcompressed = false;
+    bool rowcompressed = false;
+
+//The following refer to the current input file:
+    offset_t fileBaseOffset = 0;
+    StringAttr logicalFilename;
+    unsigned filePart = 0;
+};
+
+
+DiskRowReader::DiskRowReader(unsigned _expectedCrc, IOutputMetaData & _expected, unsigned _projectedCrc, IOutputMetaData & _projected, unsigned _actualCrc, IOutputMetaData & _actual, IPropertyTree * options)
+: expectedCrc(_expectedCrc), expectedDiskMeta(&_expected), projectedCrc(_projectedCrc), projectedDiskMeta(&_projected), actualCrc(_actualCrc), actualDiskMeta(&_actual)
+{
+    if (options)
+    {
+        //Options contain information that is the same for each file that is being read, and potentially expensive to reconfigure.
+        if (options->hasProp("encryptionKey"))
+        {
+            encryptionKey.resetBuffer();
+            options->getPropBin("encryptionKey", encryptionKey);
+        }
+        readBufferSize = options->getPropInt("readBufferSize", defaultReadBufferSize);
+    }
+}
+
+IRawRowStream * DiskRowReader::queryRawRowStream()
+{
+    return this;
+}
+
+IAllocRowStream * DiskRowReader::queryAllocatedRowStream(IEngineRowAllocator * _outputAllocator)
+{
+    outputAllocator.set(_outputAllocator);
+    return this;
+}
+
+void DiskRowReader::clearInput()
+{
+    inputBuffer.setStream(nullptr);
+    input.clear();
+}
+
+bool metaMatches(IOutputMetaData * left, IOutputMetaData * right)
+{
+    return left == right;
+}
+
+bool DiskRowReader::matches(const char * format, bool streamRemote, unsigned _expectedCrc, IOutputMetaData & _expected, unsigned _projectedCrc, IOutputMetaData & _projected, unsigned _actualCrc, IOutputMetaData & _actual, IPropertyTree * options)
+{
+    if ((expectedCrc != _expectedCrc) || (projectedCrc != _projectedCrc) || (actualCrc != _actualCrc))
+        return false;
+
+    //Is the previous check sufficient?
+    //if (!metaMatches(expectedDiskMeta, &_expected) || !metaMatches(projectedDiskMeta, &_projected) || !metaMatches(actualDiskMeta, &_actual))
+    //    return false;
+
+    if (options->hasProp("encryptionKey"))
+    {
+        MemoryBuffer tempEncryptionKey;
+        options->getPropBin("encryptionKey", tempEncryptionKey);
+        if (!encryptionKey.matches(tempEncryptionKey))
```

Comment:
I'm not clear under what circumstances this would happen.. should it throw an exception if mismatches?